File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def findpivot(nums, left, right):
            if left > right:
                return -1
            if left == right:
                return left
            mid = left + (right- left)// 2
            
            if mid < right and nums[mid] > nums[mid + 1]:
                return mid + 1
            if mid > left and nums[mid] < nums[mid - 1]:
                return mid - 1
            if nums[left] > nums[mid]:
                return findpivot(nums, left, mid - 1)
            return findpivot(nums, mid +1, right)
        
        def bsearch(nums, left, right, target):
            while left <= right:
                mid = left + (right - left) // 2
                if nums[mid] == target:
                    return mid
                elif nums[mid] > target:
                    right = mid - 1
                else:
                    left = mid + 1
            return -1
        pivot = findpivot(nums, 0, len(nums) - 1)
        
        if pivot == -1:
            return bsearch(nums, 0, len(nums) - 1, target)
        elif nums[pivot] == target:
            return pivot
        elif nums[0] <= target:
            return bsearch(nums, 0, pivot - 1, target)
        return bsearch(nums, pivot + 1, len(nums) -1 , target)

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        
        while (left <= right){
            int mid = left + (right - left) / 2;
            
            if (nums[mid] == target){
                return mid;
            }
            // check if the left element is less than the mid element whih means that part 
            //sorted and target can be found there.
            else if( nums[left] <= nums[mid]){
                // check if the target lies in the range left and mid
                // if so right = mid - 1
                // else left = mid + 1
                if( nums[left] <= target && target < nums[mid]){
                    right = mid - 1;
                }
                else{
                    left = mid + 1;
                }
            }
            else{
                if (nums[mid] < target && target <= nums[right]){
                    left = mid + 1; 
                }else{
                    right = mid - 1;
                }
            }
        }
        return -1;
    }
}